### PR TITLE
More flexible raster pipeline

### PR DESCRIPTION
**Pull request by lukasino1214**
Current raster pipeline assumes that it has vertex and fragment shader. Now it can have only vertex shader and adding other shader stages is simpler.

### CODE REVIEW

File: include/daxa/pipeline.hpp [hunk 1]
```diff
@@ -109,6 +109,10 @@ namespace daxa
     {
         ShaderInfo vertex_shader_info = {};
         ShaderInfo fragment_shader_info = {};
+        // ShaderInfo geometry_shader_info = {}; NOTE(Lukas): Maybe enable it but throw warning when used. Warning would be enabled with validation layer
```
Comment (Ipotrick) [inline include/daxa/pipeline.hpp:112]: Why are the additional stages commented out?

File: src/impl_pipeline.cpp [hunk 2]
```diff
@@ -378,48 +364,75 @@ namespace daxa
...
+                neeeded_shader_stages++;
+            }
+
+            return true;
         };
-        vkCreateShaderModule(impl.impl_device.as<ImplDevice>()->vk_device, &shader_module_pixel, nullptr, &p_vk_shader_module);
+
+        VkShaderModule v_vk_shader_module = {};
+
+        {
```
Comment (Ipotrick) [inline src/impl_pipeline.cpp:396]: This block should be part of the lambda "create_shader_module ".

File: src/impl_pipeline.cpp [hunk 3]
```diff
@@ -378,48 +364,75 @@ namespace daxa
...
-                .flags = {},
-                .stage = VkShaderStageFlagBits::VK_SHADER_STAGE_FRAGMENT_BIT,
-                .module = p_vk_shader_module,
-                .pName = modified_info.fragment_shader_info.compile_options.entry_point.value().c_str(),
-                .pSpecializationInfo = nullptr,
-            },
+        std::vector<VkPipelineShaderStageCreateInfo> vk_pipeline_shader_stage_create_infos;
+        vk_pipeline_shader_stage_create_infos.reserve(neeeded_shader_stages);
+
+        auto create_shader_stage = [&](ShaderInfo const & shader_info, VkShaderStageFlagBits shader_stage, VkShaderModule& vk_shader_module) {
```
Comment (Ipotrick) [inline src/impl_pipeline.cpp:417]: have the vector created earlier and push_back those at the same time as the shader modules

File: src/impl_pipeline.cpp [hunk 4]
```diff
@@ -378,48 +364,75 @@ namespace daxa
         //     }
         // }
 
-        VkShaderModule v_vk_shader_module = {};
-        VkShaderModule p_vk_shader_module = {};
+        u32 neeeded_shader_stages = 0;
```
Comment (GabeRundlett) [inline src/impl_pipeline.cpp:367]: What is this for? You use it to reserve a vector below. Why don't you just append to the vector with the binary WITHIN the first lambda? No reason to pre-reserve here. If it was actually a problem, we could just count up at the beginning, but not separate the work of creating the shader module and putting it into the vector

**Review by GabeRundlett — CHANGES_REQUESTED** [review]

File: src/impl_pipeline.cpp [hunk 5]
```diff
@@ -553,8 +566,14 @@ namespace daxa
 
         DAXA_DBG_ASSERT_TRUE_M(pipeline_result == VK_SUCCESS, "failed to create graphics pipeline");
 
-        vkDestroyShaderModule(impl.impl_device.as<ImplDevice>()->vk_device, v_vk_shader_module, nullptr);
-        vkDestroyShaderModule(impl.impl_device.as<ImplDevice>()->vk_device, p_vk_shader_module, nullptr);
+        auto destroy_shader_module = [&](VkShaderModule& vk_shader_module) {
```
Comment (Ipotrick) [inline src/impl_pipeline.cpp:569]: When you add a vector for shader modules, like suggested above, you can just loop over it to destroy them, no need for lambda.

**Pull request merged**